cwlVersion: v1.0
class: CommandLineTool


requirements:
- class: InlineJavascriptRequirement
  expressionLib:
  - var get_output_prefix = function(ext) {
        ext = ext || "";
        if (inputs.output_prefix == ""){
          let root = inputs.bambai_pair.basename.split('.').slice(0,-1).join('.');
          return (root == "")?inputs.bambai_pair.basename+ext:root+ext;
        } else {
          return inputs.output_prefix;
        }
    }


hints:
- class: DockerRequirement
  dockerPull: biowardrobe2/plugin-plot-rna:v0.0.2


inputs:

  annotation_file:
    type: File
    inputBinding:
      position: 5
      prefix: "-a"
    doc: |
      TSV annotation file

  bambai_pair:
    type: File
    inputBinding:
      position: 6
      prefix: "-b"
    secondaryFiles:
      - .bai
    doc: |
      Indexed BAM + BAI files

  isoforms_file:
    type: File
    inputBinding:
      position: 7
      prefix: "-i"
    doc: |
      CSV isoforms file generated by GEEP

  stats_file:
    type: File
    inputBinding:
      position: 8
      prefix: "-s"
    doc: |
      Stat file generated by python-get-stat-rnaseq.cwl

  output_prefix:
    type: string?
    inputBinding:
      position: 9
      prefix: "-o"
      valueFrom: $(get_output_prefix("_default_"))
    default: ""
    doc: |
      Output file prefix

  pair:
    type: boolean?
    inputBinding:
      position: 10
      prefix: "-p"
    doc: |
      TRUE if paired end experiment type

  dutp:
    type: boolean?
    inputBinding:
      position: 11
      prefix: "-d"
    doc: |
      TRUE if dUTP experiment type

  threads:
    type: int?
    inputBinding:
      position: 12
      prefix: "-t"
    doc: |
      Number of threads


outputs:

  png_file:
    type: File[]
    outputBinding:
      glob: "*.png"

  coverage_file:
    type: File
    outputBinding:
      glob: "*cov.tsv"

  rpkm_file:
    type: File
    outputBinding:
      glob: "*rpkm.tsv"


baseCommand: ["plot_rna.R"]