.. SPARSE documentation master file, created by
   sphinx-quickstart on Tue Nov  7 11:59:42 2017.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

Welcome to SPARSE's documentation!
==================================

SPARSE indexes >100,000 reference genomes in public databases into hierarchical clusters and uses it to predict origins of metagenomic reads.

.. toctree::
   :maxdepth: 2
   :caption: Getting Started:
   
   installation
   toyexample
   parameter

.. toctree::
   :maxdepth: 2
   :caption: Building Databases:
   
   refseq
   representative
   custom

.. toctree::
   :maxdepth: 2
   :caption: Mapping reads:

   mash
   map
.. toctree::
   :maxdepth: 2
   :caption: Interpretations:
   
   output


Citation
--------

SPARSE has been pulished as a conference paper in RECOMB 2018.
Zhou Z., Luhmann N., Alikhan NF., Quince C., Achtman M. (2018) Accurate Reconstruction of Microbial Strains from Metagenomic Sequencing Using Representative Reference Genomes. In: Raphael B. (eds) Research in Computational Molecular Biology. RECOMB 2018. Lecture Notes in Computer Science, vol 10812. Springer, Cham. DOI: https://doi.org/10.1007/978-3-319-89929-9_15

A preprint can also be found in BioRxiv : https://www.biorxiv.org/content/early/2017/11/07/215707


Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
